stock_location_donot_mix_order/models/stock_move_line.py

```diff
@@ -0,0 +1,49 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from odoo import _, models
+from odoo.exceptions import UserError
+
+
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    def _action_done(self):
+        self._check_same_order_at_destination()
+        return super()._action_done()
+
+    def _check_same_order_at_destination(self, location=None):
+        move_lines = self.filtered(
+            lambda line: line.move_id.picking_id.picking_type_id.same_next_picking
+        )
+        for line in move_lines:
+            destination_location = location or line.location_dest_id
```

Comment:
You could group the lines by picking to not make a search for each line
```suggestion
        for picking, lines in move_lines_by_picking:
            destination_locations = location or lines.location_dest_id
```

Reply:
comment added back on the new code, marking this one as resolved